File: dsa/5.binary_search/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
import typing as t

class Solution:
    def searchRange(self, nums: t.List[int], target: int) -> t.List[int]:
        left: int = 0
        right: int = len(nums) - 1

        left_bound_ind: int = -1
        right_bound_ind: int = -1

        while left <= right:
            mid: int = int((right - left) / 2) + left
            prev_val: int = nums[mid - 1] if mid - 1 >= 0 else -pow(10, 9) - 1

            if nums[mid] == target:
                if prev_val < target:
                    left_bound_ind = mid
                    break
                else:
                    right = mid - 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        left = 0
        right = len(nums) - 1
        while left <= right:
            mid: int = int((right - left) / 2) + left
            next_val: int = nums[mid + 1] if mid + 1 < len(nums) else pow(10, 9) + 1
            if nums[mid] == target:
                if next_val > target:
                    right_bound_ind = mid
                    break
                else:
                    left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1



        return [left_bound_ind, right_bound_ind]

File: dsa/5.binary_search/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.java
package binary_search;

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int leftLower = 0;
        int rightLower = nums.length - 1;

        int leftUpper = 0;
        int rightUpper = nums.length - 1;

        int leftResult = -1;
        int rightResult = -1;

        int[] result = new int[2];

        while (leftLower <= rightLower){
            int midLower = leftLower + (rightLower-leftLower)/2;
            if (nums[midLower] == target){
                if (midLower - 1 < 0){
                    leftResult = midLower;
                    break;
                }
                rightLower = midLower - 1;
            }
            else if (nums[midLower] < target){
                if (midLower + 1 < nums.length){
                    if (nums[midLower + 1] == target) {
                        leftResult = midLower + 1;
                        break;
                    }
                }
                leftLower = midLower + 1;
            }
            else{
                rightLower = midLower - 1;
            }
        }

        while (leftUpper <= rightUpper){
            int midUpper = leftUpper + (rightUpper-leftUpper)/2;
            if (nums[midUpper] == target){
                if (midUpper + 1 >= nums.length){
                    rightResult = midUpper;
                    break;
                }
                leftUpper = midUpper + 1;
            }
            else if(nums[midUpper] > target){
                if (midUpper - 1 >= 0){
                    if (nums[midUpper - 1] == target){
                        rightResult = midUpper - 1;
                        break;
                    }
                }
                rightUpper = midUpper - 1;
            }
            else {
                leftUpper = midUpper + 1;
            }
        }

        result[0] = leftResult;
        result[1] = rightResult;
        return result;
    }
}
